File: algorithms/closest_pair/closest_pair.cpp
/* 
Adaptação do código de Edson Alves, disponível em:
https://github.com/edsomjr/TEP/blob/master/Geometria_Computacional/slides/SL-2/closest.cpp
*/ 

#include <bits/stdc++.h>

using namespace std;
using point = pair<long long int, long long int>;

#define x first
#define y second

double get_distance(const point& P, const point& Q){
    return hypot(P.x - Q.x, P.y - Q.y);
}

pair<point, point> get_closest_pair(vector<point>& ps){
    size_t N = ps.size();
    sort(ps.begin(), ps.end());

    // Assume que N > 1
    auto distance = get_distance(ps[0], ps[1]);
    auto closest = make_pair(ps[0], ps[1]);

    set<point> S;
    S.insert(point(ps[0].y, ps[0].x));
    S.insert(point(ps[1].y, ps[1].x));

    for (long long int i = 2; i < N; ++i){
        auto P = ps[i];
        auto it = S.lower_bound(point(P.y - distance, 0));

        while (it != S.end()){
            auto Q = point(it->second, it->first);

            if (Q.x < P.x - distance){
                it = S.erase(it);
                continue;
            }

            if (Q.y > P.y + distance)
                break;

            auto curr_distance = get_distance(P, Q);

            if (curr_distance < distance){
                distance = curr_distance;
                closest = make_pair(P, Q);
            }

            ++it;
        }

        S.insert(point(P.y, P.x));
    }

    return closest;
}

int main(){
    vector<point> P;
    long long int x, y = 0;

    while(cin >> x >> y){
        P.push_back({x, y});
    }

    auto closest_pair = get_closest_pair(P);

    printf("%d %d\n%d %d\n",
           closest_pair.first.first, closest_pair.first.second,
           closest_pair.second.first, closest_pair.second.second
    );

    return 0;
}

File: algorithms/closest_pair/closest_pair.py
import math
import sys
from collections import namedtuple
from bisect import bisect_left
import time


Point = namedtuple('Point', 'x y')
Pair = namedtuple('Pair', 'A B')


def get_distance(P, Q):
    return math.hypot(P.x - Q.x, P.y - Q.y)


def get_closest_pair(points):
    N = len(points)
    points.sort()

    distance = get_distance(points[0], points[1])
    closest = Pair(points[0], points[1])

    S = set()

    S.add(Point(points[0].y, points[0].x))
    S.add(Point(points[1].y, points[1].x))

    S = sorted(S)

    for i in range(2, N):
        P = points[i]
        index = bisect_left(S, Point(P.y - distance, 0))
        size_s = len(S)

        while index < size_s:
            set_point = S[index]
            Q = Point(set_point.y, set_point.x)

            if Q.x < P.x - distance:
                S.remove(set_point)
                size_s = len(S)
                continue

            if Q.y > P.y + distance:
                break

            curr_distance = get_distance(P, Q)
            if curr_distance < distance:
                distance = curr_distance
                closest = Pair(P, Q)

            index += 1

        S.append(Point(P.y, P.x))

    return closest


points = []

for line in sys.stdin:
    x, y = map(int, line.split())
    points.append(Point(x, y))

begin = time.perf_counter()

closest_pair = get_closest_pair(points)

end = time.perf_counter()
elapsed = (end - begin)

print(f"Time measured: {round(elapsed, 6)} seconds.", file=sys.stderr)

print(f"{closest_pair.A.x} {closest_pair.A.y}")
print(f"{closest_pair.B.x} {closest_pair.B.y}")
